### ISSUE TITLE
Migrate add/edit product screen to Gutenberg

### ISSUE THREAD

**Issue by mrleemon** [woocommerce/woocommerce#28419]
Considering that the Classic Editor plugin will be officially supported until December 31, 2021, do you plan to migrate the add/edit product screen to use the new Gutenberg UI? If so, when? Right now, there are not many options. Both the PluginDocumentSettingPanel and PluginSidebar components are too narrow for fitting all product meta panels.

**Comment by idea--list** [comment]
Would also like to hear more about it as it is really hard to find any information in this regard.
Searching about it always results in finding info about Gutenberg blocks that will render already recorded products on the front-end, which is a totally different thing.

Now i just tried this:

1. Began editing a new Post with the block editor and turned to code mode
2. Copied that code
3. Then created a new product and pasted the gutenberg code into the description field of the product

Checking the new product on the frontend renders all the gutenberg blocks as expected. So it seems it is really just the product editing screen that still does not make use of the new block editor.

I guess this might be the reason for it:
If you drive a shop with more than a bunch of products, you probably import/export/update product data with CSV or XLS files.
It is easy to tell the importer that the information found in a spreadsheet cell should populate the description field until that filed is just text.

Problem is that if you switch to gutenberg editor inside the description field, well then you populate it with some text that describes only some structure you DESIGNED but it still does not contain any product data. Now how would you map any product data from the spreadsheet to a design structure found in WooCommerce editor's descriprion field? How would you handle the fact, that with gutenberg editor enabled, one can predefine completely different design structure for each of the products? 

That order of diversity is simply not managable by any import scripts and IMHO that is why the product editor will never get gutenbergized.

**Comment by idea--list** [comment]
Thinking a bit further: we could take use of reusable blocks. I mean we could tell the user they need first design a template that describes the layout. We could save it as a reusable block for product data and the product description field would actually be filled by that reusable block by default. Then it would contain some standardized structure and the importer script could even recognize the structure of that reusable block, so mapping would be possible... except if the user begins to define different reusable blocks for each product that would just screw up any import process once again...

The more i think about, the less feasible it seems to me.

**Comment by MaSchVam** [comment]
I'm actually sort of surprised that there is no large discussion on this topic anywhere in this repository (Perhaps there is, and I just can't find it?).  Having direct access to the block editor on every Edit Product page, would be a huge add to the general value proposition of WooCommerce, I would think?

Scouring the web, I see people have invented somewhat hacky solutions (Like: https://dev.to/kalimahapps/enable-gutenberg-editor-in-woocommerce-466m). I don't really think approaches like these are recommended, though? 

Is there actually any official word regarding this issue?

**Comment by mrleemon** [comment]
I suppose that there is no talk about this topic here because right now there is no sane way to actually fit the product data metabox into the narrow sidebar of the block editor. This mainly depends on the Gutenberg team, and they are neither talking about this in their repository, which is really strange, as there are many plugins out there that make a heavy use of metaboxes (WooCommerce is one example) and are forced to stay in the classic editor until a solution for this issue is provided.

**Comment by MaSchVam** [comment]
I agree with you. The sidebar is a problem, and I don't think it should be utilized for the product data metaboxes. I am wondering if a better approach for WooCommerce, would be to simply situate this section under the block editor, much like how Advanced Custom Fields implements its non-block related metaboxes?

Something like this, seems like the obvious approach to me, at least. I can't see how a solution is going to be provided by/inside Gutenberg, unfortunately. It is a lot of configuration to stuff into such a small place.

That is, unless WooCommerce adopts some kind of a 'pure' block approach to how all single product pages are displayed.

**Comment by mrleemon** [comment]
I've found this very old issue on the Gutenberg repo:

https://github.com/WordPress/gutenberg/issues/1684

But the discussion did not result in any significant advances in this area.

**Comment by mentosmenno2** [comment]
I was just as surprised as others here, that when I installed WooCommerce that product editing was still using the classic editor. I didn't expect that from a plugin this size. And I was indeed so surprised there are not a lot of topics/issues regarding this feature. 

But I really wanted to try it out, so I went on the way and tried to enable it myself, to see what it does.

I've used the `use_block_editor_for_post_type` filter to convert the editor to the Gutenberg editor.
I was surprised at first, because everything seemed to work just fine. But I might have been happy too early.

The possible issues that have been told in earlier comments, I didn't have. The product meta tabs weer displayed in a meta box below the editor, not in the sidebar.

**Comment by mrleemon** [comment]
```
I've used the use_block_editor_for_post_type filter to convert the editor to the Gutenberg editor.
I was surprised at first, because everything seemed to work just fine. But I might have been happy too early.

The possible issues that have been told in earlier comments, didn't occur for me. The product meta tabs weer displayed in a meta box below the editor, not in the sidebar.
```

Yes, the product meta tabs are displayed when using the `use_block_editor_for_post_type` filter, but they are not shown using native Gutenberg components, but the old meta boxes.

**Comment by fredlim** [comment]
When enabling Gutenberg with use_block_editor_for_post_type filter, the content in wysiwyg editor in Product short description cannot be display or edit. I check the Product short description section in developer console, head and body of the Visual editor is empty, but html content can be edit in Text editor.

Is that by design or a bug?

**Comment by tdobson** [comment]
I am slightly surprised this isn't better supported, or at least easily switched to.

Here's a howto of [how to enable it](https://www.youtube.com/watch?v=xJnhrmjqhTA) if you haven't already seen.

**Comment by loulemos** [comment]
I'm surprised as well, that little snippet worked but the short product description remains in the classic editor.

**Comment by raoabidali** [comment]
Adding this code to your site (example: theme functions file) will enable gutenberg for products:
```
/**
 * enable Gutenberg support for WooCommerce Products
 */

add_filter( 'use_block_editor_for_post_type', function ( $can_edit, $post_type ) {
	if ( $post_type == 'product' ) {
		$can_edit = true;
	}

	return $can_edit;
}, 10, 2 );
```

**Comment by pixelloop** [comment]
@raoabidali 

I have been using the above code but today I discovered when it's active on a product and I go to 

1) Product Data
2) Advanced
3) Try and turn "enable reviews" on or off
4) Save
5) Check front end and refresh edit product screen and the change doesn't seem to save. 

I'm not sure if there are other WooCommerce fields that are also not being saved properly but I changed the price as a test and that does seem to work so it's a bit strange.

Can you replicate that? 

UPDATE:

I have also tested this now on another WooCommerce site and had the same result.

**Comment by Endda** [comment]
still no update on gutenberg for editing products, tags, categories, etc?

**Comment by PythonDotLand** [comment]
I'm guessing many people use Elementor and such and don't need it that much? I don't like the bloat these editors add and would love to see proper Gutenberg support in WooCommerce. It's a big issue, and I think it should be high up the TODO list, to be honest.

**Comment by sstruemph** [comment]
I'm using the snippet that @raoabidali shared and the Product Short Description field is not initializing the tinymce editor's textarea field.

**Comment by FunkyCss** [comment]
There is no any feedback from Woocommerce for this?

Its 2022 and all other platforms give to user to make their product pages to look like landing pages with more than a few bold text.

I mean we really need to make the description tab as good as it deserves, with rows, columns YT videos, img backgrounds etc.

And this should be done from Woo core, not with page builders and "hacky" snippets.

We need gutenberg on product page asap at least for the long description for now.

If there is an issue with metaboxes, we should leave them there for what is necessary, but we should allow the long description to be more easy to be designed.

**Comment by Endda** [comment]
It really is sad they refuse to give updates on this issue.

**Comment by pixelloop** [comment]
@FunkyCss 100% agree.

What's most perplexing is, WooCommerce is owned by WordPress. WordPress forced Gutenberg as a core feature, and yet while they have added WooCommerce blocks to add to pages and posts, they have completely ignored how powerful blocks would be for setting up WooCommerce single product templates and descriptions etc. 

My workaround for BeaverBuilder works well, but it's still not perfect, and I'd always favour a native solution.

**Comment by mrleemon** [comment]
Yep, it's puzzling that after almost four years of Gutenberg, there's still no decent support for complex custom post types with lots of meta fields (such as products in WooCommerce). Adding meta to the right sidebar is not feasible in most cases. And both products are built by Automattic...

**Comment by coreyworrell** [comment]
Meta fields can be setup in the sidebar and can open as “flyouts” or modals to allow more space. Several plugins (and other website builders) currently use this method and it works well. 

Example: https://docs.gravityforms.com/edit-choices-flyout/

**Comment by mrleemon** [comment]
Nice idea, but I think that hiding product meta fields behind a modal in the case of a plugin such as WooCommerce is treating these as less important than content when, in fact, the opposite is true.

**Comment by jcmatoskx** [comment]
This doesn't make any sense.
Woocommerce is owned by Wordpress.
Wordpress forces post edit to work through Gutenberg, but Woocommerce products are still edited using the classic editor?
And after so many years there is no update on this end?

Talk about Woocommerce Tech debt.

**Comment by sstruemph** [comment]
@jcmatoskx Clarification... WooCommerce is owned by Automattic which owns WordPress.com... not the open-source project.

**Comment by jcmatoskx** [comment]
> @jcmatoskx Clarification... WooCommerce is owned by Automattic which owns WordPress.com... not the open-source project.

Thank you, that is true, but feels like a bad excuse/throwing dust into someones eye for 2 reasons:
-  Aren't the same people from Automattic the main drivers of the open source project anyways? Correct me if I'm wrong, but even if they aren't, It's still a terrible excuse because:
- It's their profitable/business model project (Woocommerce) the one that has fallen behind Wordpress, and is lacking the basic core feature (Gutenberg), so sure while Woo is also open source, it is also a big part of their business model so that excuse doesn't really seem to work here.

**Comment by eriky** [comment]
So it's 2023, and we're still forced to edit product pages with the old, limited editor. Or alternatively, buy Elementor Pro and add a lot of cruft (Javascript, CSS) and extra yearly costs to our sites. What's the hold-up here? Is it that difficult to implement, or is there something else at stake?

Also, are there any downside to enabling the editor with this code below?  

As show on several sites and youtube videos:
```php
// enable gutenberg for woocommerce
function activate_gutenberg_product( $can_edit, $post_type ) {
 if ( $post_type == 'product' ) {
        $can_edit = true;
    }
    return $can_edit;
}
add_filter( 'use_block_editor_for_post_type', 'activate_gutenberg_product', 10, 2 );


// enable taxonomy fields for woocommerce with gutenberg on
function enable_taxonomy_rest( $args ) {
    $args['show_in_rest'] = true;
    return $args;
}
add_filter( 'woocommerce_taxonomy_args_product_cat', 'enable_taxonomy_rest' );
add_filter( 'woocommerce_taxonomy_args_product_tag', 'enable_taxonomy_rest' );
```

**Comment by dima9891** [comment]
Looks like in version 7.7.0 implementation as above is broken now - all blocks are locked and not editable. I had to downgrade to 7.6.1 to resolve it. Maybe someone can find a solution?

**Comment by eriky** [comment]
I too have this issue now after upgrading to 7.7.0. Were you able to downgrade without issues, @dima9891? After all, there was a DB upgrade involved..

**Comment by dima9891** [comment]
@eriky i think it's better not to upgrade in the first place :) Don't see any issues, but i think it's better to restore db to pre-upgrade state just in case.

**Comment by jcaspen160** [comment]
I find it surreal that even in the year 2023, we still have to continue with the classic editor. I have had the same problem as you and I had to go back to version 7.6 because as you say breaks and blocks the page not being able to edit it. There are many suggestions and requests from users that this problem, to edit the product with gutenberg, and that from woocommerce developers have not yet solved or have not given an alternative.  And for a code that allowed us to do so now with version 7.7 can not be used. If they do not intend to include the gutenberg editor, at least allow us to continue editing the products with the code we have been using. I hope that in the next version of woocommerce this problem will be solved.

**Comment by druideniggi** [comment]
[Hello this snippet works](https://wordpress.org/support/topic/woocommerce-product-tab-locked-version-7-7-0/url). We have already migrated many products to Gutenberg editor. I hope that the Gutenberg editor master update doesn't destroy my product descriptions!!!!!
`add_filter('woocommerce_register_post_type_product', function( $args ) {
    unset( $args['template'] );
    return $args;
});`

**Comment by reesz** [comment]
After wasting an hour of my day here's how to disable this completely unnecessary locking of the editor, as well as get rid of the default template: 
```php

function wfp_wc_filter_register_post_type_product($args)
{
    $args['template']      = null;
    $args['template_lock'] = false;
    return $args;
}

add_filter('woocommerce_register_post_type_product', wfp_wc_filter_register_post_type_product, 10, 1);
```

//e: Looks like @druideniggi was 3min faster, while I was debugging. Just to be safe I also reset the `template_lock` property.

**Comment by jcaspen160** [comment]
Hi and thank you very much for the input from [Druideniggi](https://github.com/druideniggi) and [reesz](https://github.com/reesz). 
I have been testing both codes you have provided in different positions, changing and removing code from the one I have set to be able to edit with Gutenberg the product description in version 7.6. In principle the two codes activate the edition of Blocks, the problem is that when I want to update the product or gives me JSON update error or in the case of Druigeniggi code then I get the code at the top of the product page on the front end. I have included them in the code that I am using in version 7.6 of woocommerce, I have put them as a separate function, and the truth is I do not know how to put them to work in version 7.7.
The code that I am currently using and works fine in version 7.6 is: 

```
`// enable gutenberg for woocommerce

function activate_gutenberg_product( $can_edit, $post_type ) {
 if ( $post_type == 'product' ) {
        $can_edit = true;
    }
    return $can_edit;
}
add_filter( 'use_block_editor_for_post_type', 'activate_gutenberg_product', 10, 2 );

// enable taxonomy fields for woocommerce with gutenberg on
function enable_taxonomy_rest( $args ) {
    $args['show_in_rest'] = true;
    return $args;
}
add_filter( 'woocommerce_taxonomy_args_product_cat', 'enable_taxonomy_rest' );
add_filter( 'woocommerce_taxonomy_args_product_tag', 'enable_taxonomy_rest' );

/*OPCIONAL*/
/* Activar visibilidad de productos en editor de productos WooCommerce con Gutenberg */
function register_catalog_meta_boxes() {
global $current_screen;
if ( method_exists( $current_screen, 'is_block_editor' ) && $current_screen->is_block_editor() ) {
add_meta_box( 'catalog-visibility', __( 'Visibilidad en el catálogo', 'textdomain' ), 'product_data_visibility', 'product', 'side' );
}
}
add_action( 'add_meta_boxes', 'register_catalog_meta_boxes' );
function product_data_visibility( $post ) {
$thepostid = $post->ID;
$product_object = $thepostid ? wc_get_product( $thepostid ) : new WC_Product();
$current_visibility = $product_object->get_catalog_visibility();
$current_featured = wc_bool_to_string( $product_object->get_featured() );
$visibility_options = wc_get_product_visibility_options();
?>
<div class="misc-pub-section" id="catalog-visibility">
<?php esc_html_e( 'Visibilidad en el catálogo:', 'woocommerce' ); ?>
<strong id="catalog-visibility-display">
<?php
echo isset( $visibility_options[ $current_visibility ] ) ? esc_html( $visibility_options[ $current_visibility ] ) : esc_html( $current_visibility );
if ( 'yes' === $current_featured ) {
echo ', ' . esc_html__( 'Destacado', 'woocommerce' );
}
?>
</strong>
<a href="#catalog-visibility"
class="edit-catalog-visibility hide-if-no-js"><?php esc_html_e( 'Editar', 'woocommerce' ); ?></a>
<div id="catalog-visibility-select" class="hide-if-js">
<input type="hidden" name="current_visibility" id="current_visibility"
value="<?php echo esc_attr( $current_visibility ); ?>" />
<input type="hidden" name="current_featured" id="current_featured"
value="<?php echo esc_attr( $current_featured ); ?>" />
<?php
echo '<p>' . esc_html__( 'Este ajuste determina en qué páginas de la tienda se verá el producto.', 'woocommerce' ) . '</p>';
foreach ( $visibility_options as $name => $label ) {
echo '<input type="radio" name="_visibility" id="_visibility_' . esc_attr( $name ) . '" value="' . esc_attr( $name ) . '" ' . checked( $current_visibility, $name, false ) . ' data-label="' . esc_attr( $label ) . '" /> <label for="_visibility_' . esc_attr( $name ) . '" class="selectit">' . esc_html( $label ) . '</label><br />';
}
echo '<br /><input type="checkbox" name="_featured" id="_featured" ' . checked( $current_featured, 'yes', false ) . ' /> <label for="_featured">' . esc_html__( 'Este es un producto destacado', 'woocommerce' ) . '</label><br />';
?>
<p>
<a href="#catalog-visibility"
class="save-post-visibility hide-if-no-js button"><?php esc_html_e( 'Aceptar', 'woocommerce' ); ?></a>
<a href="#catalog-visibility"
class="cancel-post-visibility hide-if-no-js"><?php esc_html_e( 'Cancelar', 'woocommerce' ); ?></a>
</p>
</div>
</div>
<?php
}
```
`
can you help me how to implement what you have written so that it works well?

Thank you very much.
Best regards.

**Comment by eriky** [comment]
This is the full snippet that allowed me to upgrade to 7.7. Use at your own risk:

```php
function activate_gutenberg_product( $can_edit, $post_type ) {
 if ( $post_type == 'product' ) {
        $can_edit = true;
    }
    return $can_edit;
}
add_filter( 'use_block_editor_for_post_type', 'activate_gutenberg_product', 10, 2 );

function wfp_wc_filter_register_post_type_product($args)
{
    $args['template']      = null;
    $args['template_lock'] = false;
    return $args;
}
add_filter('woocommerce_register_post_type_product', 'wfp_wc_filter_register_post_type_product', 10, 1);
```

**Comment by druideniggi** [comment]
Thanks eriky, the snippet works perfectly. I deleted the other snippets on this topic.

**Comment by wplaunchify** [comment]
I want to throw my hat into the ring on this, as the author of LaunchFlows plugin, which uses the same snippet as @eriky has shared to add Gutenberg layout capability quite nicely to WC product pages.

The problem, as of version 7.6 of WC is that they have added the new admin "feature" of the block editor and in so doing, they have not considered that this very filter: use_block_editor_for_post_type() will also trigger an error in the console for any product page to indicate that core/xyz blocks have already been registered.

This is acknowledged as an issue by the devs contributing to the WC "updates" being (in my opinion...prematurely) forced into the WC plugin without even giving us the option to disable them with a features checkbox (similar to Ajax and redirect to cart), which would at least prevent the latest version of WC from throwing console errors into the admin

Bottom line: someone at WC has to get together with the folks in core here to decide what to do with the double loading of core blocks if anyone is using a documented filter such as this: https://developer.wordpress.org/reference/functions/use_block_editor_for_post_type/

so that it does not immediately throw an error caused by WC "experimental features" that were rushed into the main plugin.

See this thread for the history of them adding in this WC feature that is causing the issue:
https://github.com/woocommerce/woocommerce/pull/38495

**Comment by Shogun13** [comment]
Hello,

The hack to activate Gutunberg in the product description page no longer works since the last update.

It shows the classic editor… (j'en peux plus)

`function reset_product_template( $post_type_args ) {
	if ( array_key_exists( 'template', $post_type_args ) ) {
		unset( $post_type_args['template'] );
	}

	return $post_type_args;
}
add_filter( 'woocommerce_register_post_type_product', 'reset_product_template' );

// Enable Gutenberg Blocks in Product Descriptions
function activate_gutenberg_product( $can_edit, $post_type ) {
 if ( $post_type == 'product' ) {
        $can_edit = true;
    }
    return $can_edit;
}
add_filter( 'use_block_editor_for_post_type', 'activate_gutenberg_product', 20, 2 );

// Show taxonomy for woocommerce category and tags
function enable_taxonomy_rest( $args ) {
    $args['show_in_rest'] = true;
    return $args;
}
add_filter( 'woocommerce_taxonomy_args_product_cat', 'enable_taxonomy_rest' );
add_filter( 'woocommerce_taxonomy_args_product_tag', 'enable_taxonomy_rest' );`

Any idea to solve the problem? Or Woocommerce finally add this functionality via a plugin (free or paid) ?

Thanks

**Comment by pmcpinto** [comment]
The implementation of a new product creation experience is in progress. More info in the following posts:

[Introducing a brand-new product creation experience powered by Blocks](https://developer.woo.com/2023/06/28/introducing-a-brand-new-product-creation-experience-powered-by-blocks/)

[Getting to Know Woo: Extensibility in the New Product Editor](https://developer.woo.com/2023/11/17/getting-to-know-woo-extensibility-in-the-new-product-editor/)

[Block-based Product Creation Experience](https://developer.woo.com/roadmap/block-based-product-creation-experience/)

We also have a Developer Office Hours coming soon; keep an eye on the [Woo Developer Blog](https://developer.woo.com/blog/) for updates. And follow our [GitHub discussions](https://github.com/woocommerce/woocommerce/discussions/categories/woocommerce-product-block-editor), in case you have any feedback or suggestions.
